File: src/domain/value-objects/UserRole.ts
export const UserRole = {
  SuperAdministrador: 'SUPER_ADMINISTRADOR',
  Administrador: 'ADMINISTRADOR',
  Tecnico: 'TECNICO',
} as const

export type UserRole = (typeof UserRole)[keyof typeof UserRole]

export const ALL_USER_ROLES: UserRole[] = [
  UserRole.SuperAdministrador,
  UserRole.Administrador,
  UserRole.Tecnico,
]

export const WEB_USER_ROLES: UserRole[] = [
  UserRole.SuperAdministrador,
  UserRole.Administrador,
]

export const MOBILE_USER_ROLES: UserRole[] = [
  UserRole.Administrador,
  UserRole.Tecnico,
]

export function isUserRole(value: string): value is UserRole {
  return (
    value === UserRole.SuperAdministrador ||
    value === UserRole.Administrador ||
    value === UserRole.Tecnico
  )
}

export function userRoleLabel(role: UserRole): string {
  if (role === UserRole.SuperAdministrador) return 'Super Administrador'
  if (role === UserRole.Administrador) return 'Administrador'
  return 'Técnico'
}

export function userRoleAccessHint(role: UserRole): string {
  if (role === UserRole.SuperAdministrador) return 'Solo página web'
  if (role === UserRole.Administrador) return 'Página web y aplicativo móvil'
  return 'Solo aplicativo móvil'
}

export function canManageOperationalRoles(role: UserRole): boolean {
  return role === UserRole.SuperAdministrador
}

export function canManageUsers(role: UserRole): boolean {
  return (
    role === UserRole.SuperAdministrador || role === UserRole.Administrador
  )
}

export function normalizeUserRoles(values: readonly string[]): UserRole[] {
  const unique = new Set<UserRole>()
  for (const value of values) {
    if (isUserRole(value)) unique.add(value)
  }
  return ALL_USER_ROLES.filter((role) => unique.has(role)).slice(0, 3)
}

export function primaryUserRole(roles: readonly UserRole[]): UserRole | null {
  if (roles.includes(UserRole.SuperAdministrador)) {
    return UserRole.SuperAdministrador
  }
  if (roles.includes(UserRole.Administrador)) {
    return UserRole.Administrador
  }
  if (roles.includes(UserRole.Tecnico)) return UserRole.Tecnico
  return null
}

export function assignedUserRoles(user: {
  role: UserRole
  roles?: readonly UserRole[]
}): UserRole[] {
  const fromList = normalizeUserRoles(user.roles ?? [])
  if (fromList.length > 0) return fromList
  return isUserRole(user.role) ? [user.role] : []
}

export function hasAssignedRole(
  user: { role: UserRole; roles?: readonly UserRole[] },
  role: UserRole,
): boolean {
  return assignedUserRoles(user).includes(role)
}

export function webAccessRoles(user: {
  role: UserRole
  roles?: readonly UserRole[]
}): UserRole[] {
  return assignedUserRoles(user).filter((role) =>
    WEB_USER_ROLES.includes(role),
  )
}

export function mobileAccessRoles(user: {
  role: UserRole
  roles?: readonly UserRole[]
}): UserRole[] {
  return assignedUserRoles(user).filter((role) =>
    MOBILE_USER_ROLES.includes(role),
  )
}

export const CONFIGURED_SUPER_ADMIN_EMAIL = 'amamanim@unamad.edu.pe'
export const ACTIVE_ROLE_STORAGE_KEY = 'consorcio-active-role'

File: consorcio/lib/domain/value_objects/user_role.dart
enum UserRole {
  superAdministrador,
  administrador,
  tecnico;

  static UserRole fromString(String value) {
    switch (value) {
      case 'SUPER_ADMINISTRADOR':
        return UserRole.superAdministrador;
      case 'ADMINISTRADOR':
        return UserRole.administrador;
      case 'TECNICO':
        return UserRole.tecnico;
      default:
        throw ArgumentError('Rol inválido: $value');
    }
  }

  static UserRole? tryParse(String? value) {
    if (value == null || value.isEmpty) return null;
    try {
      return UserRole.fromString(value);
    } catch (_) {
      return null;
    }
  }

  String get firestoreValue {
    switch (this) {
      case UserRole.superAdministrador:
        return 'SUPER_ADMINISTRADOR';
      case UserRole.administrador:
        return 'ADMINISTRADOR';
      case UserRole.tecnico:
        return 'TECNICO';
    }
  }

  String get label {
    switch (this) {
      case UserRole.superAdministrador:
        return 'Super Administrador';
      case UserRole.administrador:
        return 'Administrador';
      case UserRole.tecnico:
        return 'Técnico';
    }
  }

  String get accessHint {
    switch (this) {
      case UserRole.superAdministrador:
        return 'Solo página web';
      case UserRole.administrador:
        return 'Página web y aplicativo móvil';
      case UserRole.tecnico:
        return 'Solo aplicativo móvil';
    }
  }

  bool get isWeb =>
      this == UserRole.superAdministrador || this == UserRole.administrador;

  bool get isMobile =>
      this == UserRole.administrador || this == UserRole.tecnico;
}

List<UserRole> normalizeUserRoles(Iterable<dynamic> values) {
  final unique = <UserRole>{};
  for (final value in values) {
    final parsed = UserRole.tryParse('$value');
    if (parsed != null) unique.add(parsed);
  }
  return [
    UserRole.superAdministrador,
    UserRole.administrador,
    UserRole.tecnico,
  ].where(unique.contains).take(3).toList();
}

UserRole? primaryUserRole(Iterable<UserRole> roles) {
  if (roles.contains(UserRole.superAdministrador)) {
    return UserRole.superAdministrador;
  }
  if (roles.contains(UserRole.administrador)) {
    return UserRole.administrador;
  }
  if (roles.contains(UserRole.tecnico)) return UserRole.tecnico;
  return null;
}
